src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/ElementCompletionResult.cs

```diff
@@ -3,35 +3,54 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.Extensions.ObjectPool;
 
 namespace Microsoft.CodeAnalysis.Razor.Completion;
 
 internal sealed class ElementCompletionResult
 {
-    public IReadOnlyDictionary<string, IEnumerable<TagHelperDescriptor>> Completions { get; }
+    private static readonly Dictionary<IEqualityComparer<string>, ObjectPool<ImmutableSortedDictionary<string, ImmutableArray<TagHelperDescriptor>>.Builder>> s_poolCache = [];
 
-    private ElementCompletionResult(IReadOnlyDictionary<string, IEnumerable<TagHelperDescriptor>> completions)
+    public ImmutableSortedDictionary<string, ImmutableArray<TagHelperDescriptor>> Completions { get; }
```

Comment:
I really don't love moving to `ImmutableSortedDictionary<,>`.  This is absolutely going to be slower for lookup over `Dicitonary<,>`. Given that, I think the pull request title is a false premise without a benchmark to show an improvement.